### ISSUE TITLE
RPG Rocket appears even with no ammo

### ISSUE THREAD

**Issue by Nyemse** [Nyemse/SPEARHEAD#219]

**Comment by scriptdotlua** [comment]
### Summary
Two issues for this one actually:

1. Rocket appearing on holster
2. Rocket appearing on weaponrig model

I've written some code to extensibly solve the first issue. The code can probably be used to further mod and add other functionality. Changes touch:
- A new module under `SPH_Weapon`
- A few lines of code in `SPH_Server`

The second issue can be solved by changing the `CheckTool` function. Not sure if this has further-reaching consequences yet, but at a first glance it doesn't seem to pose any issues.

### Issue 1:
1. Create `WeaponHooks` modulescript under RPG-7/SPH_Weapon
2. Add code (see bottom of post Annex 1 for code block)
3. Inside of `SPH_Server`, go to function `EquipGun` and underneath the call to `SetupGun`, paste:
```lua
		-- Callback invoke
		local WeaponHooks = tool:FindFirstChild("SPH_Weapon"):FindFirstChild("WeaponHooks")
		if WeaponHooks then
			local HookModule = require(WeaponHooks)
			if HookModule and HookModule.ServerHooks and HookModule.ServerHooks.Equipped then
				HookModule.ServerHooks.Equipped(tool)
			end
		end
```

### Issue 2:
Edit the function `CheckTool` in `SPH_Server` to match:
TODO: Test remaining system functionality to make sure there's no problems.
```lua
local function CheckTool(player,tool)
	if tool:FindFirstChild("SPH_Weapon") and assets.WeaponModels:FindFirstChild(tool.Name) then
		local wepStats = require(tool.SPH_Weapon.WeaponStats)
		SetupGun(tool,wepStats)
		
		CheckHolster(player,tool)
	end
end
```

### Annex 1: WeaponHooks code for RPG-7
```lua
local WeaponHooks = {}

--------------------------------------------------

WeaponHooks.ServerHooks = {}
WeaponHooks.ClientHooks = {}

--------------------------------------------------

local function HandleWeaponDisplay(WeaponTool : Tool)
	local AmmoFolder : Folder = WeaponTool:FindFirstChild("Ammo")
	if not AmmoFolder then warn("Could not find ammo folder for weapon",WeaponTool,"in EquippingCallback") return end

	local MagAmmoValue : DoubleConstrainedValue = AmmoFolder:FindFirstChild("MagAmmo")
	if not MagAmmoValue then return end

	local CurrentMagAmmo = MagAmmoValue.Value

	--

	local Character = WeaponTool.Parent
	if not Character:FindFirstChildWhichIsA("Humanoid") then return end

	local WeaponRig = Character:FindFirstChild("WeaponRig")
	if not WeaponRig then return end

	local WeaponModel = WeaponRig:FindFirstChild("Weapon"):FindFirstChild(WeaponTool.Name)
	if not WeaponModel then warn("Could not find weapon model for weapon",WeaponTool,"in EquippingCallback") return end

	--

	local RocketPart : MeshPart = WeaponModel:FindFirstChild("Rocket")
	if not RocketPart then return end

	local RocketMesh : MeshPart = RocketPart:FindFirstChild("Rocket")
	if not RocketMesh then return end

	if CurrentMagAmmo == 0 then
		RocketMesh.Transparency = 1
	elseif CurrentMagAmmo == 1 then
		RocketMesh.Transparency = 0
	else
		warn("WARN: Rocket has 1+ Ammo!")
	end
end

--TODO: Potentially look into passing the weapon module as well
-- Convention: Equipping = fires before the weapon is equipped (head call)
-- Convention: Equipped  = fires after the weapon is equipped (tail call)
WeaponHooks.ServerHooks.Equipped = function(WeaponTool : Tool)
	
	HandleWeaponDisplay(WeaponTool)
	
end

--------------------------------------------------

return WeaponHooks
```

**Comment by Nyemse** [comment]
@scriptdotlua This seems like a pretty complicated solution to what I assumed was just some faulty if-statements, is this all really necessary for this to work? The information needed to check if the rocket should be visible or not should already be available in the Tool's attributes or ammo values server-side, so it should just be matter of reading those values and changing the rocket visibility.

**Comment by scriptdotlua** [comment]
I wrote it like this to make further modifications to the system cleaner, instead of having to edit the main system logic. **You're right** -- it could be solved easily with an if statement, but I believe this way is more extensible.

For example, what if you have a multi-part projectile? You would need to add a for loop and pollute the main server code, whereas with this you could just add a hook definition. And what if you wanted more complex, weapon-specific behaviors? It wouldn't make sense to add non-general cases to the main code.

**Comment by Maxxxxxx-x** [comment]
The only multipart projectiles tested on firearms are duplex / triplex rounds, which are scrapped as they aren't effective enough to be used by the general public. 

The only multipart projectiles i can think of are APFSDS rounds for tanks, HVP for naval and a few types of missiles or missiles with MRV warheads. 

Given that the project's primary focus are firearms, I don't think adding a whole function hook feature to SPEARHEAD which mainly targets multipart projectiles / complex weapon behavior would be needed to solve edge case an if statement can fix? Especially when, as the comments from the snippets you provided, the code isn't tested?

**Comment by scriptdotlua** [comment]
The functionality works as I am using it in my own fork of Spearhead. As I said, that's (multipart projectiles) only a contrived example. I'm not creative enough to come up with one on the spot right now.

The hook functionality wouldn't solely/mainly target multipart projectiles, it just so happens that the projectile show/hide system is something the system could solve.

The hook system doesn't introduce any performance overhead, and the main bulk of the code is bundled together with the weapon the hook is made for (which makes it more maintainable). The hook system itself would just be adding ~7 lines of code to wherever we want to fire a hook event. This could be put into a function like `FireHook("Equipped", tool)`.

Either way it helps decouple the projectile hiding functionality from the main code, which was my main point. Just because an issue can be solved simply (using an if statement), does not mean that it is a proper solution.

**Comment by scriptdotlua** [comment]
But either way, if we choose not to add the hooks I don't mind. I provided a fix to the other issue above.

**Comment by Nyemse** [comment]
I agree that SPH needs something like this hook system to allow more custom code to be written for guns, however this is completely overkill for a minor visual bug. This would be more relevant to something like animations or hit effects.

**Comment by goldfinder** [comment]
I can look into the rocket issue.

**Comment by Nyemse** [comment]
The SpawnGun and EquipGun functions in SPH_Server were missing the code needed to check for projectile visibility, adding a couple lines to check if the current gun needs to have projectile visibility changed seems to have fixed the issue.
